pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+}
+
+func (h *v3Helper) GuaranteedUpdate(ctx context.Context, key string, ptrToType runtime.Object, ignoreNotFound bool, tryUpdate storage.UpdateFunc) error {
+	panic("TODO: unimplemented")
+}
+
+func (h *v3Helper) Watch(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("TODO: unimplemented")
+}
+
+func (h *v3Helper) WatchList(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("TODO: unimplemented")
+}
+
+func keyWithPrefix(prefix, key string) string {
+	if strings.HasPrefix(key, prefix) {
+		return key
+	}
+	return path.Join(prefix, key)
+}
+
+// extractRevision extracts revision from obj.
+// If obj has resource version >0, it will be reset to 0.
+func extractRevision(versioner storage.Versioner, obj runtime.Object) (int64, error) {
+	rev, err := versioner.ObjectResourceVersion(obj)
+	if err != nil {
+		return 0, errors.New("couldn't get resourceVersion from object")
+	}
+	if rev != 0 {
+		// We cannot store object with resourceVersion in etcd, we need to clear it here.
+		if err := versioner.UpdateObject(obj, nil, 0); err != nil {
+			return 0, errors.New("resourceVersion cannot be set on objects store in etcd")
+		}
+	}
+	return int64(rev), nil
+}
+
+// decode decodes value of bytes into object. It will also set the object resource version to rev.
+// On success, objPtr would be set to the object.
+func decode(codec runtime.Codec, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error {
+	if _, err := conversion.EnforcePtr(objPtr); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+	_, _, err := codec.Decode(value, nil, objPtr)
+	if err != nil {
+		return err
+	}
+	// being unable to set the version does not prevent the object from being extracted
+	versioner.UpdateObject(objPtr, nil, uint64(rev))
+	return nil
+}
+
+// decodeList decodes a list of values into a list of objects, with resource version set to corresponding rev.
+// On success, ListPtr would be set to the list of objects.
+func decodeList(values [][]byte, revs []uint64, filter storage.FilterFunc, ListPtr interface{}, codec runtime.Codec, versioner storage.Versioner) error {
+	v, err := conversion.EnforcePtr(ListPtr)
+	if err != nil || v.Kind() != reflect.Slice {
+		panic("need ptr to slice")
+	}
+	for i, val := range values {
+		obj, _, err := codec.Decode(val, nil, reflect.New(v.Type().Elem()).Interface().(runtime.Object))
+		if err != nil {
+			return err
+		}
+		// being unable to set the version does not prevent the object from being extracted
+		versioner.UpdateObject(obj, nil, revs[i])
+		if filter(obj) {
+			v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+		}
+	}
+	return nil
+}
+
+// setupRoutine would do some common routine.
+func setupRoutine(ctx context.Context, key, prefix string) (context.Context, string) {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
```

Comment:
Agreed. Elide this check all together and force the caller to provide one, even if it's just `context.TODO()`. It's better to have the TODOs at the calling sites than taken care of here.